// stylelint/dist/rules/selector-pseudo-element-no-unknown/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.messages = exports.ruleName = undefined;

exports.default = function (actual, options) {
  return function (root, result) {
    var validOptions = (0, _utils.validateOptions)(result, ruleName, { actual: actual }, {
      actual: options,
      possible: {
        ignorePseudoElements: [_lodash.isString]
      },
      optional: true
    });
    if (!validOptions) {
      return;
    }

    root.walkRules(function (rule) {
      var selector = rule.selector;

      if (selector.indexOf("::") === -1) {
        return;
      }

      (0, _postcssSelectorParser2.default)(function (selectorTree) {
        selectorTree.walkPseudos(function (pseudoNode) {
          var pseudoElement = pseudoNode.value;

          // Ignore pseudo-classes
          if (pseudoElement.indexOf("::") === -1) {
            return;
          }

          var pseudoElementName = pseudoElement.replace(/:+/, "");

          if (_postcss.vendor.prefix(pseudoElementName) || (0, _utils.isKnownPseudoElement)(pseudoElementName)) {
            return;
          }

          var ignorePseudoElements = options && options.ignorePseudoElements || [];

          if (ignorePseudoElements.indexOf(pseudoElementName) !== -1) {
            return;
          }

          (0, _utils.report)({
            message: messages.rejected(pseudoElement),
            node: rule,
            index: pseudoNode.sourceIndex,
            ruleName: ruleName,
            result: result
          });
        });
      }).process(selector);
    });
  };
};

var _lodash = require("lodash");

var _postcssSelectorParser = require("postcss-selector-parser");

var _postcssSelectorParser2 = _interopRequireDefault(_postcssSelectorParser);

var _postcss = require("postcss");

var _utils = require("../../utils");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var ruleName = exports.ruleName = "selector-pseudo-element-no-unknown";

var messages = exports.messages = (0, _utils.ruleMessages)(ruleName, {
  rejected: function rejected(u) {
    return "Unexpected unknown pseudo-element selector \"" + u + "\"";
  }
});
